device-side heap experiment ("slab")

froystig opened the pull request:
Our compiler stack (roughly) requires that array shapes be static at compile time, which allows it to statically schedule computation and memory for all intermediate arrays. How could we lift this requirement? How would programs that lower with some shape dynamism fare practically in comparison to alternatives, like a collection of many static-shape programs, or a hand-written Pallas kernel? This introduces an experiment meant to answer questions like these.

Specifically, this adds:

* The `slab` module, which implements a heap-like data structure on device, backed by one big (statically shaped) array that we call a "slab." It exposes functions to dynamically allocate arrays on the slab, and basic operations over these dynamic-shape arrays, such as `matmul`. The operations work by iterating a static-shape computation across tiles of the dynamic-shape array. The internal shape calculations (needed for e.g. de/allocation and tiling) are carried out as integer operations on the device. Everything is compatible with `jax.jit`.
* The `djax` module, which stands up a simple "dynamic shape jax" frontend (the function `djit`) and lowers it to static-shape jax by evaluation using `slab` operations.

Work with the inimitable @apaszke and @mattjj!

froystig committed:
heap/slab dynamic shape machine experiment

Basic support for elementwise ops and f32, to start.

froystig committed:
new slab layout

Namely:

* The slab is now `n` by 512 bytes, each row considered a
  "word". Addressing is word-aligned for now.
* Program values are now assumed 1-dimensional, for the time being.

Also:

* Tile shapes are now parameters of the tiling template operation.
* Views carry dtypes. The slab is stored in `uint8` and bitcast
  according to a view as needed.

Co-authored-by: Adam Paszke <apaszke@google.com>
Co-authored-by: Matthew Johnson <mattjj@google.com>

froystig committed:
extend elementwise operations to multi-dimensional views

Co-authored-by: Matthew Johnson <mattjj@google.com>

froystig committed:
rewrite the slab in the units that we (think we) want

Addressing is now in phrases.

Co-authored-by: Adam Paszke <apaszke@google.com>

froystig committed:
reintroduce elementwise operations

Also:
* basic binop "tests"
* conveniences for alloc-and-write and whole-array read
* strides fix

Co-authored-by: Adam Paszke <apaszke@google.com>

froystig committed:
implement matmul!

Bugfixes in `slab_read` and `slab_write` along the way.

Co-authored-by: Adam Paszke <apaszke@google.com>

froystig committed:
introduce a dynamic-shape jax frontend that lowers via slabby jax

Namely, stand up a lowering from dynamic-shape jaxpr ("djaxpr") to
static-shape jaxpr with slab operations (or really to XLA
altogether). The lowering proceeds by interpreting a djaxpr using the
slab library to evaluate each djaxpr primitive. The action of this
interpreter is then compiled using the standard `jax.jit`, which
specializes to the static shapes of the slab, plus input interpreter
context.

As an frontend entry point to this, introduce a higher-order function,
`djit`, that resolves abstracted axes to dimension sizes and stages
its argument to a djaxpr.

We want the `djit` function also handles allocating and copying
input/output arrays to/from the slab before/after dispatching the
compiled backend computation. At the same time, we also wanted `djit`
calls to compatible program input to `jax.jit`.  In order to do both,
we rely on constant-forwarding rules across its inner "backend" `jit`
call to calculate concrete output shapes. Specifically, we need the
forwarding improvements in https://github.com/google/jax/pull/21414.

This change also includes some cleanups along the way, including a
unified elementwise kernel template (for any arity of operation).

Co-authored-by: Matthew Johnson <mattjj@google.com>

froystig committed:
add helpers for internal static shape/dimension assertions

Also some typing fixes.

apaszke commented on jax/experimental/slab/djax.py:
```diff
@@ -0,0 +1,165 @@
+
+import jax.experimental.slab as sl
+
+jax.config.update('jax_dynamic_shapes', True)
```
Should we really have this at the top level of a file in JAX?

froystig replied:
I was using the context manager in an earlier iteration and @mattjj asked that we switch to this. Either way, turning it on ought to be fully backwards compatible in principle (right?).

mattjj replied:
Yes it should be fully backward compatible _except_ for some edge-case errors may have to be raised later than they currently are (e.g. at lowering time rather than trace time, though TBH I forget the example here).

If we factor this to be more like a module and less like a script then it would make sense to have the jax_dynamic_shapes=True setting in the test file and in the user code, rather than in the module to be imported. That is, I agree we should remove this line from any modules-to-be-imported-by-users, and I think it'll naturally get pulled into any example files / tests instead, just like the other script-like lines currently in this file.

froystig replied:
We could also go back to enabling it on the inside with dynamic scope? Our use of dynamic jaxpr staging is localized to roughly one line of code.

Tests would also have to use some kind of local/dynamic scoping, otherwise they'd be turning it on for everything in process.

froystig replied:
I brought back configuration under a dynamic scope.

It seems that enabling dynamic shapes globally (as a side effect of import) caused failures elsewhere in the test suite in CI. I didn't dig deeper, but the issue disappeared once I implemented things under a dynamic scope and removed the global config update.

apaszke commented on jax/experimental/slab/slab.py:
```diff
@@ -0,0 +1,367 @@
+def print_seg(msg):
+  print()
+  print(f'-- {msg}')
+  print()
```
This is probably primarily for testing? Perhaps remove it?

apaszke commented on jax/experimental/slab/slab.py:
```diff
@@ -0,0 +1,367 @@
+
+
+if __name__ == '__main__':
+  main(sys.argv[1:])
```
Should we just add a new test file?

froystig replied:
Yeah, I should factor out into tests in both modules.

apaszke commented on jax/experimental/slab/djax.py:
```diff
@@ -0,0 +1,165 @@
+
+
+if __name__ == '__main__':
+  main(sys.argv[1:])
```
Consider adding a test file for this as well?

apaszke reviewed (approved):
LGTM, but it might be worth structuring it more like a JAX feature and less as a script

froystig committed:
remove usage of `eval_shape` when defining allocating ops

This was unnecessarily involved and introduced an unenforced
requirement that shape-dynamism be enabled during the evaluation.

froystig committed:
configure dynamic shape mode only locally

... under a dynamic scope when staging out djaxprs. This avoids
enabling dynamic shape mode globally (in turn also making it a side
effect of `djax` import).
